File: 第七章_二叉树/二叉树(4)/树上最远距离练习题/LongestDistance.java
/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;
    public TreeNode(int val) {
        this.val = val;
    }
}*/
public class LongestDistance {
	public int findLongest(TreeNode root) {
		int l_maxdis[] = new int[1];// 左子树跨节点的最大长度
		int ld = getDepth(root.left, l_maxdis);// 左子树的最大深度
		int r_maxdis[] = new int[1];// 右子树跨节点的最大长度
		int rd = getDepth(root.right, r_maxdis);// 右子树的最大深度

		return Math.max(ld + rd + 1, Math.max(l_maxdis[0], r_maxdis[0]));
	}

	public int getDepth(TreeNode root, int maxDistance[]) {
		if (root == null) {
			return 0;
		}

		int l = getDepth(root.left, maxDistance);
		int r = getDepth(root.right, maxDistance);

		maxDistance[0] = l + r + 1;// 跨节点的最大长度

		return Math.max(l, r) + 1;
	}
}

File: 第七章_二叉树/二叉树(4)/树上最远距离练习题/LongestDistance.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};*/
 
class LongestDistance {
public:
    int findLongest(TreeNode* root) {
        
       int lmax1=0;// 左子树跨节点的最大长度
       int lres=getMaxHeight(root->left,lmax1);// 左子树的最大深度
     
       int rmax1=0;// 右子树跨节点的最大长度
       int rres=getMaxHeight(root->right,rmax1);// 右子树的最大深度
         
             
       return max(lres+rres+1,max(lmax1,rmax1));
    }
     
 
    int max(int a,int b)
    {
        return a>b?a:b;
    }
     
    int getMaxHeight(TreeNode* root,int &maxDistance)
    {
        if(root==NULL)
        {
            return 0;
        }
         
         
        int lh=getMaxHeight(root->left,maxDistance);
        int rh=getMaxHeight(root->right,maxDistance);
 
        maxDistance=lh+rh+1;// 跨节点的最大长度
        return max(lh,rh)+1;
    }
};;